external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCSpout.java

```diff
@@ -0,0 +1,530 @@
+import org.netpreserve.jwarc.MessageBody;
+import org.netpreserve.jwarc.ParsingException;
+import org.netpreserve.jwarc.WarcPayload;
+import org.netpreserve.jwarc.WarcReader;
+import org.netpreserve.jwarc.WarcRecord;
+import org.netpreserve.jwarc.WarcRequest;
+import org.netpreserve.jwarc.WarcResponse;
+import org.netpreserve.jwarc.WarcTruncationReason;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.digitalpebble.stormcrawler.Constants;
+import com.digitalpebble.stormcrawler.Metadata;
+import com.digitalpebble.stormcrawler.persistence.Status;
+import com.digitalpebble.stormcrawler.protocol.ProtocolResponse;
+import com.digitalpebble.stormcrawler.spout.FileSpout;
+import com.digitalpebble.stormcrawler.util.ConfUtils;
+
+/**
+ * Read WARC files from the local files system and emit the WARC captures as
+ * tuples into the topology same way as done by
+ * {@link com.digitalpebble.stormcrawler.bolt.FetcherBolt}.
+ */
+@SuppressWarnings("serial")
+public class WARCSpout extends FileSpout {
+
+    private static final Logger LOG = LoggerFactory.getLogger(WARCSpout.class);
+
+    private int maxContentSize = -1;
+    private int contentBufferSize = 8192;
+    private long sleepEmitFetched = 0;
+
+    private boolean storeHTTPHeaders = false;
+    private String protocolMDprefix = "";
+
+    private WarcReader warcReader;
+    private String warcFileInProgress;
+    private WarcRequest precedingWarcRequest;
+    private Optional<WarcRecord> record;
+
+    private MultiCountMetric eventCounter;
+
+    public WARCSpout(String... files) {
+        super(false, files);
+    }
+
+    public WARCSpout(String dir, String filter) {
+        super(dir, filter, false);
+    }
+
+    /**
+     * Holder of truncation status when WARC payload exceeding the content
+     * length limit (http.content.limit) is truncated.
+     */
+    public static class TruncationStatus {
+        boolean isTruncated = false;
+        int originalSize = -1;
+
+        public void set(boolean isTruncated) {
+            this.isTruncated = isTruncated;
+        }
+
+        public boolean get() {
+            return isTruncated;
+        }
+
+        public void setOriginalSize(int size) {
+            originalSize = size;
+        }
+
+        public int getOriginalSize() {
+            return originalSize;
+        }
+    }
+
+    private void openWARC() {
+        if (warcReader != null) {
+            try {
+                warcReader.close();
+            } catch (IOException e) {
+                LOG.warn("Failed to close open WARC file", e);
+            }
+            warcReader = null;
+        }
+
+        byte[] head = buffer.removeFirst();
+        List<Object> fields = _scheme.deserialize(ByteBuffer.wrap(head));
+        warcFileInProgress = (String) fields.get(0);
+        if (warcFileInProgress == null)
+            return;
+
+        LOG.info("Reading WARC file {}", warcFileInProgress);
+        ReadableByteChannel warcChannel = null;
```

Comment:
try with resource to simplify the code and remove the close section below?

Reply:
Good point. Thanks!

Reply:
Actually, it does not work: the channel needs to stay, it'll be closed together by `warcReader.close()`.

Reply:
ok!